components/places/src/error.rs

```diff
@@ -163,3 +163,158 @@ pub enum InvalidMetadataObservation {
     #[error("Observed view time is invalid (too long)")]
     ViewTimeTooLong,
 }
+
+#[derive(Debug, thiserror::Error)]
+pub enum PlacesError {
+    #[error("Unexpected error: {0}")]
+    UnexpectedPlacesException(String),
+
+    #[error("UrlParseFailed: {0}")]
+    UrlParseFailed(String),
+
+    #[error("PlacesConnectionBusy error: {0}")]
+    PlacesConnectionBusy(String),
+
+    #[error("Operation Interrupted: {0}")]
+    OperationInterrupted(String),
+
+    /**
+     * Error indicating bookmarks corruption. If this occurs, we
+     * would appreciate reports.
+     *
+     * Eventually it should be fixed up, when detected as part of
+     * `runMaintenance`.
+     */
+    #[error("BookmarksCorruption error: {0}")]
+    BookmarksCorruption(String),
+
+    /**
+     * Thrown when providing a guid referring to a non-folder as the
+     * parentGUID parameter to a create or update
+     */
+    #[error("Invalid Parent: {0}")]
+    InvalidParent(String),
+
+    /**
+     * Thrown when providing a guid to a create or update function
+     * which does not refer to a known bookmark.
+     */
+    #[error("Unknown bookmark: {0}")]
+    UnknownBookmarkItem(String),
+
+    /**
+     * Thrown when attempting to insert a URL greater than 65536 bytes
+     * (after punycoding and percent encoding).
+     *
+     * Attempting to truncate the URL is difficult and subtle, and
+     * is guaranteed to result in a URL different from the one the
+     * user attempted to bookmark, and so an error is thrown instead.
+     */
+    #[error("URL too long: {0}")]
+    UrlTooLong(String),
+
+    /**
+     * Thrown when attempting to update a bookmark item in an illegal
+     * way. For example, attempting to change the URL of a bookmark
+     * folder, or update the title of a separator, etc.
+     */
+    #[error("Invalid Bookmark: {0}")]
+    InvalidBookmarkUpdate(String),
+
+    /**
+     * Thrown when:
+     *
+     * - Attempting to insert a child under BookmarkRoot.Root,
+     * - Attempting to update any of the bookmark roots.
+     * - Attempting to delete any of the bookmark roots.
+     */
+    #[error("CannotUpdateRoot error: {0}")]
+    CannotUpdateRoot(String),
+
+    #[error("Unexpected error: {0}")]
+    InternalPanic(String),
+}
+
+// A port of the error conversion stuff that was in ffi.rs - it turns our
+// "internal" errors into "public" ones.
+impl From<Error> for PlacesError {
+    fn from(error: Error) -> PlacesError {
+        let label = error.to_string();
+        let kind = error.kind();
+        match kind {
+            ErrorKind::InvalidPlaceInfo(info) => {
+                log::error!("Invalid place info: {}", info);
```

Comment:
I'm not sure at all how log here works, but will this hit sentry? (and do we want it to?)

Reply:
Not entirely sure, would have to go a little deeper to figure out where this eventually logs to. This was mainly a port from ffi.rs so I assume it's doing whatever it was previously.

Reply:
Yeah, that does seem a little heavy-handed - all error matches [here](https://github.com/mozilla/application-services/blob/574450d9f3c1f5c9c6c2a71345cf0678440e985f/components/places/src/ffi.rs#L199) log an error, and we can see them in sentry - eg, loads of `places::ffi - URL parse error` and surprisingly only 1 `Invalid place info:`.

FWIW, these errors end up in sentry via [this helper](https://github.com/mozilla-mobile/android-components/blob/6ba529f9538b8ea9ffba1846821a76cc80c3de17/components/support/rustlog/src/main/java/mozilla/components/support/rustlog/RustLog.kt#L106)

I think we might as well keep things as they are for now just to reduce the moving parts in this patch, but do agree with the implication we should think this through a bit better (eg, all the "URL parse error" entries are probably fairly useless and probably just reflect a user typing something in the awesomebar)

Reply:
(and just for fun<sup>1</sup>, check out our ["unexpected error"s](https://sentry.prod.mozaws.net/operations/firefox/?query=is%3Aunresolved+%22unexpected+error%22) 🙄 )

<sup>1</sup> May not contain fun

Reply:
See also [where a-c submits these to sentry](https://github.com/mozilla-mobile/android-components/blob/0b5b051a6751077902784a235404a577c158b9e3/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt#L92) - so I *think* that many of these exceptions end up being submitted twice.

Reply:
> See also [where a-c submits these to sentry](https://github.com/mozilla-mobile/android-components/blob/0b5b051a6751077902784a235404a577c158b9e3/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt#L92) - so I _think_ that many of these exceptions end up being submitted twice.

Hmm interesting, so this would technically be an issue that is existing in today's world? Since the ffi.rs currently has that `log:error` for the old code while this would handle errors for any of the uniffi-ed errors? This may be something I'll need to also dive into to wrap my head around.

Reply:
Sorry I wasn't clear, but yes, this is a comment on the existing state of the world and nothing to do with your patch. It's possible that sentry is smart enough to combine these (eg, there will be an exact partial stack match when both are submitted, which might be enough for sentry to know they are the same)

Just yesterday I created https://github.com/mozilla-mobile/android-components/pull/11295 to try and avoid some of these errors ending up in sentry, but it turns out that might not actually work given this code *will* still report them. We probably do need to step back and have a think about this in more general terms.

Reply:
ah totally makes sense now -- I think your patch will definitely lessen the noise of things we won't/shouldn't fix and definitely rethink just converting these errors and continuing sending up to sentry (like in the case of this patch)